abd-clam/src/cakes/mod.rs

```diff
@@ -1,6 +1,340 @@
 //! CLAM-Accelerated K-nearest-neighbor Entropy-scaling Search.
 
+use core::ops::Index;
+
 pub mod knn;
 pub mod rnn;
-pub mod sharded;
-pub mod singular;
+mod search;
+mod sharded;
+mod singular;
+
+use distances::Number;
+use rayon::prelude::*;
+use search::Search;
+use sharded::RandomlySharded;
+use singular::SingleShard;
+
+use crate::{Dataset, Instance, PartitionCriteria};
+
+/// CAKES search.
+pub enum Cakes<I: Instance, U: Number, D: Dataset<I, U>> {
+    /// Search with a single shard.
+    SingleShard(SingleShard<I, U, D>),
+    /// Search with multiple shards.
+    RandomlySharded(RandomlySharded<I, U, D>),
+}
+
+impl<I: Instance, U: Number, D: Dataset<I, U>> Cakes<I, U, D> {
+    /// Creates a new CAKES instance with a single shard dataset.
+    ///
+    /// # Arguments
+    ///
+    /// * `data` - The dataset to search.
+    /// * `seed` - The seed to use for the random number generator.
+    /// * `criteria` - The criteria to use for partitioning the tree.
+    pub fn new(data: D, seed: Option<u64>, criteria: &PartitionCriteria<U>) -> Self {
+        Self::SingleShard(SingleShard::new(data, seed, criteria))
+    }
+
+    /// Creates a new CAKES instance with a randomly sharded dataset.
+    ///
+    /// # Arguments
+    ///
+    /// * `shards` - The shards of the dataset to search.
+    /// * `seed` - The seed to use for the random number generator.
+    /// * `criteria` - The criteria to use for partitioning the tree.
+    #[must_use]
```

Comment:
Why is this here?

Reply:
Clippy

Reply:
The idea is that a pure function (one with no side effects such as I/O) is pointless if its return value isn't used, so Clippy suggests you put in that attribute. That attribute causes the compiler to issue a warning if you ever call the function without using its return value. It does NOT mean that code must call that function.